### PR TITLE
[bug](jsonb) fix jsonb query bug When the json key value contains "." 

### PR DESCRIPTION
**Pull request by LemonLiTree**
# Proposed changes

Issue Number: close https://github.com/apache/doris/issues/19173

mysql> SELECT jsonb_extract('{"a.b.c":{"k1":"v31", "k2.a1": 300},"a":"opentelemetry"}', '$."a.b.c".k1');
+-------------------------------------------------------------------------------------------+
| jsonb_extract('{"a.b.c":{"k1":"v31", "k2.a1": 300},"a":"opentelemetry"}', '$."a.b.c".k1') |
+-------------------------------------------------------------------------------------------+
| "v31"                                                                                     |
+-------------------------------------------------------------------------------------------+
1 row in set (0.06 sec)

### CODE REVIEW

File: be/src/util/jsonb_document.h [hunk 1]
```diff
@@ -1099,16 +1099,38 @@ inline JsonbValue* JsonbValue::findPath(const char* key_path, unsigned int kp_le
...
+                        key_has_delim = true;
+                    }
+                }
             }
         }
 
+        if (key_has_delim) {
+            key++;
+            klen = klen - 2;
+        }
```
Comment (gitccl) [inline be/src/util/jsonb_document.h:1132]: Maybe we can judge `if (left_quotation_marks != nullptr && right_quotation_marks != nullptr)` to support case `$."a"`?

**Comment by LemonLiTree** [comment]
run buildall

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.65 seconds
 stream load tsv:          426 seconds loaded 74807831229 Bytes, about 167 MB/s
 stream load json:         24 seconds loaded 2358488459 Bytes, about 93 MB/s
 stream load orc:          59 seconds loaded 1101869774 Bytes, about 17 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230430100637_clickbench_pr_137443.html

**Review by zy-kkk — DISMISSED** [review]

File: be/src/util/jsonb_document.h [hunk 2]
```diff
@@ -1099,16 +1099,33 @@ inline JsonbValue* JsonbValue::findPath(const char* key_path, unsigned int kp_le
...
         size_t idx_len = 0;
         // find the current key and [] bracket position
-        for (; key_path != fence && *key_path != *delim; ++key_path, ++klen) {
+        for (; key_path != fence; ++key_path, ++klen) {
             if ('[' == *key_path) {
                 left_bracket = key_path;
             } else if (']' == *key_path) {
                 right_bracket = key_path;
+            } else if ('"' == *key_path) {
+                if (left_quotation_marks == nullptr) {
```
Comment (xiaokang) [inline be/src/util/jsonb_document.h:1112]: should consider more complicated cases:
1. unclosed double quote in the key, eg. 'a".b' -> a . "b
2. there are many double quote pairs in the key, eg. '"a.b"c".d"' -> a.bc.d
3. there is double quote escape in the key, eg. 'a".b.\"c"' -> a.b".c

**Comment by LemonLiTree** [comment]
run buildall

**Comment by LemonLiTree** [comment]
run buildall

File: be/src/util/jsonb_document.h [hunk 3]
```diff
@@ -213,6 +226,83 @@ class JsonbDocument {
...
+    void appendLeg(char a) { leg += a; }
+
+    void clearLeg() { leg.clear(); }
+
+    void setLeg(std::string a) {
+        clearLeg();
+        leg = a;
+    }
+
+    std::string getLeg() { return leg; }
```
Comment (xiaokang) [inline be/src/util/jsonb_document.h:284]: It's may be slow since it copies string

File: be/src/util/jsonb_document.h [hunk 4]
```diff
@@ -213,6 +226,83 @@ class JsonbDocument {
...
+        m_position = std::find_if_not(m_position, m_end, [](char c) { return std::isspace(c); });
+    }
+
+    /// Moves the position n bytes forward.
+    void skip(size_t n) {
+        assert(remaining() >= n);
+        m_position += n;
+    }
+
+    void appendLeg(char a) { leg += a; }
```
Comment (xiaokang) [inline be/src/util/jsonb_document.h:275]: It's not efficient to append string leg.

**Comment by LemonLiTree** [comment]
run buildall

File: be/src/util/jsonb_document.h [hunk 5]
```diff
@@ -1072,103 +1186,165 @@ inline const char* JsonbValue::getValuePtr() const {
...
-            // check the last char is ]
-            if (key + klen - 1 != right_bracket) {
-                return nullptr;
+
+            pval = ((ObjectVal*)pval)->find(stream.get_legPtr(), stream.get_legLen(), handler);
+
+            if (!pval) return nullptr;
+        } else if (LIKELY(pval->type_ == JsonbType::T_Array)) {
+            int index = 0;
+            std::string idx_string(stream.get_legPtr(), stream.get_legLen());
```
Comment (xiaokang) [inline be/src/util/jsonb_document.h:1223]: try to use string_view to avoid memory copy

File: be/src/util/jsonb_document.h [hunk 6]
```diff
@@ -1072,103 +1186,165 @@ inline const char* JsonbValue::getValuePtr() const {
...
+
+            if (stream.get_legLen() == 1 && *stream.get_legPtr() == WILDCARD) {
+                return pval;
+            } else if (std::string(stream.get_legPtr(), 4) == LAST) {
+                auto pos = idx_string.find(MINUS);
```
Comment (xiaokang) [inline be/src/util/jsonb_document.h:1228]: Is it possible since == LAST

File: be/src/util/jsonb_document.h [hunk 7]
```diff
@@ -1072,103 +1186,165 @@ inline const char* JsonbValue::getValuePtr() const {
...
+                    }
+                    index = num - 1 - std::stoi(idx_string);
+                } else if (stream.get_legLen() == 4) {
+                    index = ((ArrayVal*)pval)->numElem() - 1;
+                } else {
+                    return nullptr; //invalid json path
+                }
+            } else {
+                std::string::size_type pos;
+                index = std::stoi(idx_string, &pos, 10);
```
Comment (xiaokang) [inline be/src/util/jsonb_document.h:1244]: handle negative index here.

File: be/src/util/jsonb_document.h [hunk 8]
```diff
@@ -213,6 +226,109 @@ class JsonbDocument {
...
+
+private:
+    /// The current position in the stream.
+    const char* m_position;
+
+    /// The end of the stream.
+    const char* const m_end;
+
+    ///path leg ptr
+    char* legPtr;
```
Comment (xiaokang) [inline be/src/util/jsonb_document.h:314]: leg_ptr

File: be/src/util/jsonb_document.h [hunk 9]
```diff
@@ -213,6 +226,109 @@ class JsonbDocument {
...
+
+    ///path leg len
+    unsigned int legLen;
```
Comment (xiaokang) [inline be/src/util/jsonb_document.h:317]: leg_len

File: be/src/util/jsonb_document.h [hunk 10]
```diff
@@ -213,6 +226,109 @@ class JsonbDocument {
...
+
+    ///
+    bool hasEscapes = false;
```
Comment (xiaokang) [inline be/src/util/jsonb_document.h:320]: has_escapes

**Review by xiaokang — APPROVED** [review]
LGTM

**Review by yiguolei — APPROVED** [review]
LGTM

**Pull request merged**